### PR TITLE
Infra spell bot

### LINKED ISSUE #12522: Add US English spell checker/linter for markdown
We should have a US English spell checker. 

https://github.com/mdn/content/issues/373 changed all know UK English spellings to US English, as per MDN policy. However this is likely to rot. As discussed in #10787 we should also have an automated spell checker system in place for English content.

### PR DESCRIPTION
**Pull request by OnkarRuikar**
- fixes https://github.com/mdn/content/issues/12522

I've been running a [spell-checker bot](https://github.com/OnkarRuikar/temp/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aclosed) in my `temp` repo for over 1.5 years. The bot runs every Monday start and files an issue giving [details about found typos/spelling](https://github.com/OnkarRuikar/temp/issues/69) mistakes in the `mdn/content` repo. The issue is resolved by fixing typos in `mdn/content` and adding new words in the `project-words.txt` file.

In the last weekly meeting, we got the green signal to move the bot to the content repo. The PR moves the bot to this repo.

The following changes have been made to existing files:
- The `project-words.txt` file contains 5k+ ignored words gathered over the period.
- The RegExs have been updated to handle umlauts in fragments, to allow `valid=true` case
- There are multiple cases of `favourite-colour`. As the article is named in such a way, we have to ignore all the occurrences.

### CODE REVIEW

**Review by rebloor — APPROVED** [review]
For the corrections to favourite-colour

File: .github/workflows/spelling-check-bot.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        uses: dacbd/create-issue-action@main
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          title: Weekly spelling check
+          body: |
+            ## Found typos and unknown words:
+            ${{ env.OUTPUT }}
+
+            > [!TIP]
+            > Add valid words to the https://github.com/mdn/content/blob/main/.vscode/project-words.txt file, and fix the typos.
```
[the remark this answers is not in the document]
Comment (OnkarRuikar) [reply .github/workflows/spelling-check-bot.yml:49]: I hope the typo issues will not linger more than a day. A maintainer won't even have to wait for a reviewer.

File: .vscode/project-words.txt
Comment (bsmth) [inline .vscode/project-words.txt:1]: One thing that would be great to check is if `ignoreWords` can accept an external file https://cspell.org/configuration/

The rationale is that we're registering the project-words as a CSpell dictionary so gibberish words from here will get suggested for VSCode autocorrect, which is not ideal.

Comment (OnkarRuikar) [reply .vscode/project-words.txt:1]: :thinking: We can use a separate config and wordlist file for the workflow:
```
output=$(npx cspell --no-progress --config workflow-cspell.json "content/**/*.md" || exit 0)
```

**Comment by Josh-Cena** [comment]
@OnkarRuikar Perhaps you can split out the content change into a separate PR so it can be quickly merged? https://github.com/mdn/content/issues/35404

**Comment by Josh-Cena** [comment]
@OnkarRuikar Could you split all content changes into separate PRs? We don't want to loop more and more people into this.

File: .vscode/project-words.txt
Comment (OnkarRuikar) [reply .vscode/project-words.txt:1]: Good idea! The [`noSuggest`](https://github.com/streetsidesoftware/cspell/blob/39401d46cc54b3715721ed171849c80a1c9b7990/packages/cspell-eslint-plugin/assets/options.schema.json#L118) option in dictionary definition does what you want. I've created a separate `ignored-words` dictionary using `ignored-words.txt` file.

Now `project-words.txt` is to hold actual technology terms like [`mmultiscripts`](https://developer.mozilla.org/en-US/docs/Web/MathML/Element/mmultiscripts); the worlds which we would like to be used as suggestions in the IDEs.  And `ignored-words.txt` will hold custom terms like `mycolor`, `Caniuse`, `onkar`, `bsmith`, etc.

**Comment by Josh-Cena** [comment]
We already have `project-words.txt` right? Why two?

**Comment by OnkarRuikar** [comment]
> However I think in ignored-words.txt some of them are real words like "granularities". This is going to be tricky ;)

Because the bot was working standalone and the editor consideration was not there so all the words were accumulated in one file. There are 5578 words in the file, and classifying them is a huge time-consuming task. But we can do it later after the bot goes live.


If we could force contributors to use camel case variable names then a ton of words will get removed. :roll_eyes:   `myvar` vs `myVar`

**Comment by Josh-Cena** [comment]
> If we could force contributors to use camel case variable names then a ton of words will get removed. 🙄

I would not be opposed to a PR that goes slightly beyond fixing real typos and also reduces the dictionary size! In practice contributors won't be "forced" to do anything anyway since the typo check is out of band.

File: .vscode/project-words.txt
Comment (bsmth) [reply .vscode/project-words.txt:1]: >Now project-words.txt is to hold actual technology terms like [mmultiscripts](https://developer.mozilla.org/en-US/docs/Web/MathML/Element/mmultiscripts); the words which we would like to be used as suggestions in the IDEs. And ignored-words.txt will hold custom terms like mycolor, Caniuse, onkar, bsmith, etc.

I like this a lot better, thanks for researching

File: .github/workflows/spelling-check-bot.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  workflow_dispatch:
+
+jobs:
+  sync:
+    if: github.repository == 'mdn/content'
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+
```
Comment (bsmth) [inline .github/workflows/spelling-check-bot.yml:15]: We have this step twice, is that right? Should we move the full checkout step here and delete the one below?

```suggestion
      - uses: actions/checkout@v4
        with:
          repository: mdn/content
          ref: main
          path: content
```

File: .github/workflows/spelling-check-bot.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        with:
+          node-version-file: ".nvmrc"
+          cache: yarn
+
+      - uses: actions/checkout@v4
+        with:
+          repository: mdn/content
+          ref: main
+          path: content
+
```
Comment (OnkarRuikar) [inline .github/workflows/spelling-check-bot.yml:27]:
```suggestion
```

File: .github/workflows/spelling-check-bot.yml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  sync:
+    if: github.repository == 'mdn/content'
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          repository: mdn/content
+          ref: main
+          path: content
```
Comment (OnkarRuikar) [inline .github/workflows/spelling-check-bot.yml:18]:
```suggestion
          path: content

```

File: .github/workflows/spelling-check-bot.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        uses: dacbd/create-issue-action@main
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          title: Weekly spelling check
+          body: |
+            ## Found typos and unknown words:
+            ${{ env.OUTPUT }}
+
+            > [!TIP]
+            > Add valid words to the [ignored-words.txt](https://github.com/mdn/content/blob/main/.vscode/ignored-words.txt) or [project-words.txt](https://github.com/mdn/content/blob/main/.vscode/project-words.txt) file, and fix the typos.
```
Comment (bsmth) [inline .github/workflows/spelling-check-bot.yml:47]:
```suggestion
            > To exclude words from the spellchecker, you can add valid words (web technology terms or abbreviations) to the [project-words.txt](https://github.com/mdn/content/blob/main/.vscode/project-words.txt) dictionary for IDE autocompletion. To ignore strings that are not words (`AABBCC` in code, for instance), you can add them to [ignored-words.txt](https://github.com/mdn/content/blob/main/.vscode/ignored-words.txt).
```

File: .github/workflows/spelling-check-bot.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+      - name: Create an issue
+        if: env.OUTPUT != ''
+        uses: dacbd/create-issue-action@main
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          title: Weekly spelling check
+          body: |
+            ## Found typos and unknown words:
+            ${{ env.OUTPUT }}
```
Comment (bsmth) [inline .github/workflows/spelling-check-bot.yml:44]: Does this make sense to put it in a code block?

```suggestion
            Typos and unknown words:
            
            ```
            ${{ env.OUTPUT }}
            ```
```

**Comment by OnkarRuikar** [comment]
I've successfully tested the workflow.
Created Issue: https://github.com/OnkarRuikar/content/issues/29
The workflow run: https://github.com/OnkarRuikar/content/actions/runs/10591216755/job/29348284353#step:4:45

**Comment by bsmth** [comment]
> I've successfully tested the workflow. Created Issue: [OnkarRuikar#29](https://github.com/OnkarRuikar/content/issues/29) The workflow run: https://github.com/OnkarRuikar/content/actions/runs/10591216755/job/29348284353#step:4:45

It looks good, tnx. I think we can merge this shortly. I would also echo the sentiment from Josh that we should try to reduce the 5k+ size dictionary by some means in a follow-up. Like:

* moving actual technology terms from ignore-words to project-words
* camelCasing variable names (requires content updates)
* Some project documentation about the spellchecker, how to add words, what kinds of words go where (`CONTRIBUTING.md`?)

**Review by bsmth — CHANGES_REQUESTED** [review]
One last thing I noticed; we have dupes in the `ignore-words.txt` file, could you `uniq` it first? 

__edit:__ Also if you think renaming the dictionaries will help with their purpose, we can do that in this PR, for instance `ignored-words` -> `ignore-list`?

**Comment by OnkarRuikar** [comment]
I ran `sort file | uniq file > file` on the word list files. How about names `project-words-list.txt` and `ignored-words-list.txt` for the files?

**Comment by bsmth** [comment]
> I ran `sort file | uniq file > file` on the word list files. 

nice, thank you

> How about names `project-words-list.txt` and `ignored-words-list.txt` for the files?

IMO "words" is the part to get rid of in the ignore file, so I'd prefer to keep the original or do something like:

```
terms-abbreviations.txt
ignore-list.txt
```

What do you reckon?

**Comment by OnkarRuikar** [comment]
@bsmth `terms-abbreviations.txt` and `ignore-list.txt` sound good I've renamed the files and updated rest of the content.

**Review by bsmth — APPROVED** [review]
Looking good, let's give it a shot! Thanks, Onkar

**Comment by OnkarRuikar** [comment]
Don't forget to trigger the workflow manually.

**Pull request merged**